### PR TITLE
add support for iterative build of OpenBLAS with 64-bit integers

### PR DESCRIPTION
**Pull request by jfgrimm**
(created using `eb --new-pr`)

### CODE REVIEW

**Comment by akesandgren** [comment]
I don't think adding _ilp64 suffix to the actual symbols is a good thing. Most codes I know of that does use ilp64 blas just assumes that the symbols are the same as usual.
I know I never did that when compiling it with 64bit integers way back and the codes that relied on it worked just fine that way

**Comment by jfgrimm** [comment]
@akesandgren good point, I've switched back to using `LIBNAMESUFFIX`

**Comment by migueldiascosta** [comment]
I agree it's simpler not to suffix the symbols, but otoh, numpy suggests using them to avoid clashes (https://numpy.org/devdocs/user/building.html#bit-blas-and-lapack), and SciPy says they actually need both (https://scipy.github.io/devdocs/dev/contributor/building.html#reference-for-build-options)... 

Could we build all three variants?

**Comment by jfgrimm** [comment]
@migueldiascosta so, two 64bit iterations: one iteration with `INTERFACE64=1 SYMBOLSUFFIX=64_` and one with `INTERFACE64=1 SYMBOLSUFFIX=ilp64` ?

**Comment by migueldiascosta** [comment]
overriding the symbols is `SYMBOLSUFFIX=` (empty), I think

**Comment by akesandgren** [comment]
Conflict resolution needed.

Also, I think we should try to get both the plain 32bit int and the 64_ symbol suffix functions into the same libopenblas.(a|so) library in case that isn't already part of this change.

**Comment by bartoldeman** [comment]
I did the conflict resolution. There are a few things to note however:
* There is probably a case to be made for 3 libraries:
  1. `libopenblas.so` (LP64)
  2. `libopenblas64.so` (ILP64, no suffixes). This naming while ambiguous matches Fedora and Debian's pre-existing practice. This library can be easily used by FlexiBLAS via `libflexiblas64.so`  (that's the default naming in FlexiBLAS, too), and then by ScaLAPACK, and various others (nwchem/globalarrays e.g.)
  3. `libopenblas_ilp64.so` (or something else, with suffixes). This won't work with FlexiBLAS.

Precompiled Julia uses case 3, as `libopenblas.so_`. Since Julia can dlopen random other libraries, it makes sense *not* to combine ILP64 symbols and LP64 symbols into a single library as that would cause a conflict with Julia then.

There are also the CBLAS names, which confuse me a little so far; simply using `SYMBOLSUFFIX=64_` gives you `cblas_dgemm64_()` (it is that way in Julia's OpenBLAS), where `cblas_dgemm_64()` would make more sense. There is some functionality in CBLAS added since this PR was opened, I need to double check if OpenBLAS supports that now.

**Comment by jfgrimm** [comment]
@bartoldeman is that documentation up to date? I'm ending up with `libopenblas_64.so`, not `libopenblas64.so`. 
Also, reference lapack has already implemented the `_64` suffix before fortran compiler mangling: https://github.com/Reference-LAPACK/lapack/blob/master/CMakeLists.txt#L119

**Comment by bartoldeman** [comment]
@jfgrimm the _64 suffix isn't supported by OpenBLAS yet (it uses its own implementation of cblas that is not in sync with reference lapack). It will take some time for the ecosystem to catch up (FlexiBLAS and BLIS also)

the underscore is a little nasty as it was automatically added until this commit:
https://github.com/OpenMathLib/OpenBLAS/commit/ca6b4961e4a5d280#diff-45268af9f7d4b41382cdaae2c9d9b4ff7e773f3cdfccff4a29b107cc955f4cd0

the Debian approach is probably better, see 
https://salsa.debian.org/science-team/openblas/-/blob/master/debian/rules?ref_type=heads
basically, using `LIBPREFIX=libopenblas` or `LIBPREFIX=libopenblas64` instead works with all recent OpenBLAS versions.
I can work on that.

**Comment by jfgrimm** [comment]
Yes, I was just looking at the debian recipe. We'll want to fix the .pc file to match as well

**Comment by bartoldeman** [comment]
done in 
https://github.com/jfgrimm/easybuild-easyblocks/pull/5

**Comment by jfgrimm** [comment]
Test report by @jfgrimm
#### Overview of tested easyconfigs (in order)
 * **SUCCESS** _OpenBLAS-0.3.15-GCC-10.3.0.eb_ 
 * **SUCCESS** _OpenBLAS-0.3.20-GCC-11.3.0.eb_ 
 * **FAIL (build issue)** _OpenBLAS-0.3.23-GCC-12.3.0.eb_ (partial log available at https://gist.github.com/jfgrimm/f285f0506783946a5b0d6b841d8b6b6d)
 * **SUCCESS** _OpenBLAS-0.3.27-GCC-13.3.0.eb_ 

Build succeeded for 3 out of 4 (4 easyconfigs in total)
node097.viking2.yor.alces.network - Linux Rocky Linux 8.9, x86_64, AMD EPYC 7643 48-Core Processor, Python 3.6.8
See https://gist.github.com/jfgrimm/c962f1c40abd55f6475d6be41c9fddd9 for a full test report.

**Comment by bartoldeman** [comment]
Test report by @bartoldeman
#### Overview of tested easyconfigs (in order)
 * **SUCCESS** _OpenBLAS-0.3.23-GCC-12.3.0.eb_ 

Build succeeded for 1 out of 1 (1 easyconfigs in total)
login1.int.archimedes.c3.ca - Linux AlmaLinux 9.4 (Seafoam Ocelot), x86_64, Intel Xeon Processor (Skylake, IBRS) (cascadelake), Python 3.11.4
See https://gist.github.com/bartoldeman/8a93a10ea4838ff44977704ba346c62f for a full test report.

**Comment by bartoldeman** [comment]
Test report by @bartoldeman
#### Overview of tested easyconfigs (in order)
 * **SUCCESS** _OpenBLAS-0.3.27-GCC-13.3.0.eb_ 

Build succeeded for 1 out of 1 (1 easyconfigs in total)
login1.int.archimedes.c3.ca - Linux AlmaLinux 9.4 (Seafoam Ocelot), x86_64, Intel Xeon Processor (Skylake, IBRS) (cascadelake), Python 3.11.4
See https://gist.github.com/bartoldeman/f63533c8bf71de7e4073a16957a157ce for a full test report.

**Comment by bartoldeman** [comment]
Test report by @bartoldeman
#### Overview of tested easyconfigs (in order)
 * **SUCCESS** _OpenBLAS-0.3.29-GCC-14.2.0.eb_ 

Build succeeded for 1 out of 1 (1 easyconfigs in total)
login1.int.archimedes.c3.ca - Linux AlmaLinux 9.4 (Seafoam Ocelot), x86_64, Intel Xeon Processor (Skylake, IBRS) (cascadelake), Python 3.11.4
See https://gist.github.com/bartoldeman/5eb510c8bfbd594b394a3e1b11a22cd9 for a full test report.

**Comment by bartoldeman** [comment]
@jfgrimm not sure why your 0.3.23 failed..
we'll have to deal somehow with easyconfigs such as `OpenBLAS-0.3.27-GCC-13.3.0-seq-iface64.eb`. They compile fine, just generate some extra libraries where the `libopenblas64.{so,a}` libraries are identical to the `libopenblas.{so,a}` libraries, it looks backwards compatible to me.

**Comment by jfgrimm** [comment]
Could be some local hooks or something interfering... I'm on leave for a week starting today, so won't have an opportunity to check for a while.

**Comment by bartoldeman** [comment]
Test report by @bartoldeman
#### Overview of tested easyconfigs (in order)
 * **SUCCESS** _OpenBLAS-0.3.27-GCC-13.3.0-seq-iface64.eb_ 

Build succeeded for 1 out of 1 (1 easyconfigs in total)
login1.int.archimedes.c3.ca - Linux AlmaLinux 9.4 (Seafoam Ocelot), x86_64, Intel Xeon Processor (Skylake, IBRS) (cascadelake), Python 3.11.4
See https://gist.github.com/bartoldeman/ca689f4d92decedd38ba389db0321d05 for a full test report.

**Comment by bartoldeman** [comment]
@jfgrimm any chance of rechecking your failed `OpenBLAS-0.3.23-GCC-12.3.0.eb`?

**Comment by jfgrimm** [comment]
@bartoldeman will retry in a sec

**Comment by jfgrimm** [comment]
Test report by @jfgrimm
#### Overview of tested easyconfigs (in order)
 * **FAIL (build issue)** _OpenBLAS-0.3.23-GCC-12.3.0.eb_ (partial log available at https://gist.github.com/jfgrimm/e2bd3008bf13a8b1e950d3650f78ab0d)

Build succeeded for 0 out of 1 (1 easyconfigs in total)
node034.viking2.yor.alces.network - Linux Rocky Linux 8.9, x86_64, AMD EPYC 7643 48-Core Processor, Python 3.6.8
See https://gist.github.com/jfgrimm/b1f98472e326f258cd69b79f3434ada9 for a full test report.

**Comment by jfgrimm** [comment]
@bartoldeman I get a ton of undefined lapack/lapacke references in `linktest`:
```
 119 gcc -O2 -DSMALL_MATRIX_OPT -DMAX_STACK_ALLOC=2048 -fopenmp -Wall -m64 -DF_INTERFACE_GFORT -fPIC -DSMP_SERVER -DUSE_OPENMP -DNO_WARMUP -DMAX_CPU_NUMBER=96 -DMAX_PARALLEL_NUMBER=1 -DBUILD     _SINGLE=1 -DBUILD_DOUBLE=1 -DBUILD_COMPLEX=1 -DBUILD_COMPLEX16=1 -DVERSION=\"0.3.23\" -msse3 -mssse3 -msse4.1 -mavx -mavx2 -mavx2 -UASMNAME -UASMFNAME -UNAME -UCNAME -UCHAR_NAME -UCHAR_     CNAME -DASMNAME= -DASMFNAME=_ -DNAME=_ -DCNAME= -DCHAR_NAME=\"_\" -DCHAR_CNAME=\"\" -DNO_AFFINITY -I.. -L/opt/apps/eb/software/Python/3.11.3-GCCcore-12.3.0/lib64 -L/opt/apps/eb/software     /Python/3.11.3-GCCcore-12.3.0/lib -L/opt/apps/eb/software/GCCcore/12.3.0/lib64 -L/opt/apps/eb/software/GCCcore/12.3.0/lib -w -o linktest linktest.c ../libopenblas_zenp-r0.3.23.so -L/opt     /apps/eb/software/Python/3.11.3-GCCcore-12.3.0/lib/../lib64 -L/opt/apps/eb/software/OpenSSL/1.1/lib/../lib64 -L/opt/apps/eb/software/libffi/3.4.4-GCCcore-12.3.0/lib64/../lib64 -L/opt/ap     ps/eb/software/libffi/3.4.4-GCCcore-12.3.0/lib/../lib64 -L/opt/apps/eb/software/XZ/5.4.2-GCCcore-12.3.0/lib/../lib64 -L/opt/apps/eb/software/SQLite/3.42.0-GCCcore-12.3.0/lib/../lib64 -L     /opt/apps/eb/software/Tcl/8.6.13-GCCcore-12.3.0/lib/../lib64 -L/opt/apps/eb/software/libreadline/8.2-GCCcore-12.3.0/lib/../lib64 -L/opt/apps/eb/software/ncurses/6.4-GCCcore-12.3.0/lib/.     ./lib64 -L/opt/apps/eb/software/bzip2/1.0.8-GCCcore-12.3.0/lib/../lib64 -L/opt/apps/eb/software/binutils/2.40-GCCcore-12.3.0/lib/../lib64 -L/opt/apps/eb/software/zlib/1.2.13-GCCcore-12.     3.0/lib/../lib64 -L/opt/apps/eb/software/GCCcore/12.3.0/lib/gcc/x86_64-pc-linux-gnu/12.3.0 -L/opt/apps/eb/software/GCCcore/12.3.0/lib/gcc/x86_64-pc-linux-gnu/12.3.0/../../../../lib64 -L     /lib/../lib64 -L/usr/lib/../lib64 -L/opt/apps/eb/software/Python/3.11.3-GCCcore-12.3.0/lib -L/opt/apps/eb/software/OpenSSL/1.1/lib -L/opt/apps/eb/software/libffi/3.4.4-GCCcore-12.3.0/li     b64 -L/opt/apps/eb/software/libffi/3.4.4-GCCcore-12.3.0/lib -L/opt/apps/eb/software/XZ/5.4.2-GCCcore-12.3.0/lib -L/opt/apps/eb/software/SQLite/3.42.0-GCCcore-12.3.0/lib -L/opt/apps/eb/s     oftware/Tcl/8.6.13-GCCcore-12.3.0/lib -L/opt/apps/eb/software/libreadline/8.2-GCCcore-12.3.0/lib -L/opt/apps/eb/software/ncurses/6.4-GCCcore-12.3.0/lib -L/opt/apps/eb/software/bzip2/1.0     .8-GCCcore-12.3.0/lib -L/opt/apps/eb/software/binutils/2.40-GCCcore-12.3.0/lib -L/opt/apps/eb/software/zlib/1.2.13-GCCcore-12.3.0/lib -L/opt/apps/eb/software/GCCcore/12.3.0/lib/gcc/x86_     64-pc-linux-gnu/12.3.0/../../..  -lgfortran -lm -lgomp -lquadmath -lm -lpthread -lc  && echo OK.
 120 /opt/apps/eb/software/binutils/2.40-GCCcore-12.3.0/bin/ld: /tmp/eb-w2p9ij_5/ccFZq0qF.o: in function `main':
 121 linktest.c:(.text.startup+0xf88): undefined reference to `slagge_'
 122 /opt/apps/eb/software/binutils/2.40-GCCcore-12.3.0/bin/ld: linktest.c:(.text.startup+0xf8f): undefined reference to `slagsy_'
 123 /opt/apps/eb/software/binutils/2.40-GCCcore-12.3.0/bin/ld: linktest.c:(.text.startup+0xf96): undefined reference to `slahilb_'
[...]
```

**Comment by jfgrimm** [comment]
Test report by @jfgrimm
#### Overview of tested easyconfigs (in order)
 * **SUCCESS** _OpenBLAS-0.3.23-GCC-12.3.0.eb_ 

Build succeeded for 1 out of 1 (1 easyconfigs in total)
node034.viking2.yor.alces.network - Linux Rocky Linux 8.9, x86_64, AMD EPYC 7643 48-Core Processor, Python 3.6.8
See https://gist.github.com/jfgrimm/66cd3e9426642fdbbbbd434691199650 for a full test report.

**Comment by jfgrimm** [comment]
@bartoldeman success! also, fixed the minor merge conflict

**Review by bartoldeman — APPROVED** [review]
LGTM
merging now that EB 5.1 has been released
